examples/access-control-migration/src/services/assign-project-instance-id.voter.ts

```diff
@@ -0,0 +1,35 @@
+import {
+  AuthorizationContext,
+  AuthorizationMetadata,
+  AuthorizationDecision,
+} from '@loopback/authorization';
+import _ from 'lodash';
+
+// Instance level authorizer
+// This is a WORKAROUND to modify the authorization context
+// It is not used for making a decision, so just returns ABSTAIN
+export async function assignProjectInstanceId(
+  authorizationCtx: AuthorizationContext,
+  metadata: AuthorizationMetadata,
+) {
+  // hard-coded for a collection of known endpoints
+  // TBD(FEAT): We should support instance level context for object,
+  // like project1, project2, project3
+  // the current AuthrozationCtx is designed for class level object,
+  // like projects
+  const projectId = authorizationCtx.invocationContext.args[0];
+  const resourceName = getResourceName(metadata.resource ?? authorizationCtx.resource, projectId);
+  // resourceId will override the resource name from metadata
+  authorizationCtx.resourceId = resourceName;
+  return AuthorizationDecision.ABSTAIN;
+}
+
+// TBD: REFACTOR to a casbin util file
+// Generate the resource name according to the naming convention
+// in casbin policy
+function getResourceName(resource: string, id?: number): string {
```

Comment:
if we plan on having `u1` in the casbin files, then we should have a table that maps `u1` to a unique `email` or `userid` instead of a database `index`.  [securityId] and id in principal are usually the same and are database uuid or indexes. I think for casbin, this will cause problems .

Reply:
not sure I understand:
>  [securityId] and id in principal are usually the same and are database uuid or indexes.
That's why I create the user field in casbin with conversion `u${id}`, to ensure it's unique.
So what is the problem here?

Reply:
What I am saying is that right now, the database id for a user is : 1,2,3, etc. But if I store stuff in Mongodb it will be a long uuid or guid, so you would need `u0209b175-b39c-4dd4-a158-4dc81547e77e` in your casbin file. And if ever users are deleted and added to a database again, these guid values will be different, and so you will need to change your casbin files accordingly.

So having users in memory database:

```
   "User": {
      "1": "{\"id\":1,\"username\":\"John\",\"email\":\"john@doe.com\"}",
      "2": "{\"id\":2,\"username\":\"Jane\",\"email\":\"jane@doe.com\"}",
      "3": "{\"id\":3,\"username\":\"Bob\",\"email\":\"bob@projects.com\"}"
    },
```

and casbin file using  `u1`, `u2` etc  works for this example, but perhaps a more real world example would place
unique user names in the casbin file like: `jannyhou88` and `dremond71` , and in real database these users would have ID values like a guid or long int.

So what I am saying is that we should have a function that takes the securityId and queries the database for the unique userid, and this userid is used by cabin policy decisions.

Reply:
@emonddr Hmm, this is a design choice that I don't quite agree:
> the database id for a user is : 1,2,3, etc. But if I store stuff in Mongodb it will be a long uuid or guid, so you would need u0209b175-b39c-4dd4-a158-4dc81547e77e in your casbin file.

First, my app as a prototype demo for RBAC, wants to avoid advise on those noisy implementation details not related to the core authorization process.
The design principle I aim to emphasize is [what to do](https://github.com/strongloop/loopback-next/blob/363251f2069b4586afc8668d3eb5f315aed6be46/examples/access-control-migration/imgs/authorization-casbin.png): what authorization system does, what 3rd-party lib does. The diagram explaining that concepts is added in the doc PR, NOT "how to do".

So, things like "how principal(user) id maps to the 3rd-party lib's record"(what we are discussing here), "how to define your casbin policy models", "how to optimize the model relations", etc...really depends on each project's business requirements. IMO An example app shouldn't be over engineered to advise on such decisions.

Back to your question, "But if I store stuff in Mongodb it will be a long uuid or guid...", since my app is created with memory db, I don't have this concern and I don't see any value of designing an example in a way that it is able to support all popular databases. Like I use `opensesame` as the password for everyone to avoid dealing with designing a robust security system.

And in a real project, I don't know is it a good idea to have the back-end data coupled with the 3rd-party library. For example, what if I decide to not use casbin but use oauth0 instead? What if I change the casbin model and the subject uses a combination of 2 fields in the user model? This is also a reason why I'd like to make the back-end data agnostic of the 3rd-party lib.

WDYT?

Reply:
Ok. We can leave it as-is for the first version. :)

Reply:
thanks there will be a few follow-up stories after this PR, will definitely include our discussion here. 🤝